### PR TITLE
feat: i18n

### LINKED ISSUE #29: [Feature] Enable i18n/l10n Support
Current statically defined labels, strings, char/relic/LC, etc. are hard-coded en-US.

To enable international support, refactor app, DB, components, etc. to support locale-based display.

These changes may necessitate a UI/UX pass as non-latin languages will take less/more horizontal space in tables, cards, etc.

### PR DESCRIPTION
**Pull request by FI00ds**
# Pull Request

## Description

* enables translation across the website
* Only adds some basic translations for non-english (generated from game data)

## Related Issue

* Closes #29 

## Checklist

- [x] Initial Translation support
  - [x] Language selector
  - [x] Modals
    - [x] Scoring
    - [x] 0 results
    - [x] many perms
    - [x] 0 perms
    - [x] edit character
    - [x] relic
    - [x] edit image
    - [x] save build
    - [x] switch relics
    - [x] builds
    - [x] score footer
    - [x] select character
    - [x] select lightcone
  - [x] Menu Drawer
  - [x] Relic Scorer Tab
  - [x] Optimizer Tab
    - [x] Sidebar
    - [x] Form
      - [x] Character Options
      - [x] Set conditionals
      - [x] Filters
      - [x] Teammates
      - [x] Simulation
    - [x] Grid
    - [x] Translation support for Content / Teammate Content
      - [x] Lightcones
        - [x] 5*
        - [x] 4*
        - [x] 3*
      - [x] Characters
        - [x] 5*
        - [x] 4*
  - [x] Characters Tab
    - [x] Character Preview
      - [x] Relic Preview
      - [x] Stat Score
      - [x] Combat Score
        - [x] Selection Sliders
        - [x] dmg upgrades
        - [x] combat stats
        - [x] build analysis
        - [x] scoring summary
    - [x] Grid
    - [x] Filter Bar
    - [x] Character Actions
  - [x] Relics Tab
    - [x] Filter Bar
    - [x] Relic Insights
    - [x] Grid
  - [x] Import/Save Tab
  - [x] Partial import messages
  - [x] Settings
  - [x] Get Started Tab
    - [x] Language specific images support
  - [x] Hints
  - [x] Add en text as comments to improve legibility
  - [x] Break DPS scoring explanation into multiple sections
  - [x] Write i18n readme
  - [x] Add missing Conditionals
    - [x] DanceAtSunset
    - [x] IVentureForthToHunt
    - [x] ScentAloneStaysTrue
    - [x] ThoseManySprings
    - [x] PoisedToBloom
    - [x] ShadowedByNight

### CODE REVIEW

File: src/components/CharacterModal.jsx [hunk 1]
```diff
@@ -52,25 +55,25 @@ export default function CharacterModal(props) {
...
       footer={[
-        <Button key="back" onClick={handleCancel}>
-          Cancel
+        <Button key='back' onClick={handleCancel}>
+          {t('cancel')}
         </Button>,
-        <Button key="submit" type="primary" onClick={onModalOk}>
-          Save
+        <Button key='submit' type='primary' onClick={onModalOk}>
+          {t('save')}
```
Comment (fribbels) [inline src/components/CharacterModal.jsx:62]: Is there a reason these strings are t('lowercase') instead of t('Uppercase')? Feels like this might get confusing at some point

File: src/components/RelicModal.tsx [hunk 2]
```diff
@@ -173,20 +200,27 @@ export default function RelicModal(props: {
...
       mainStatOptions = Object.entries(Constants.PartsMainStats[formValues.part]).map((entry) => ({
-        label: entry[1],
+        label: (() => {
+          return (
+            <Flex align='center' gap={10}>
+              <img src={Assets.getStatIcon(entry[1], true)} style={{ width: 22, height: 22 }}/>
+              {t(`common:stats.${entry[1]}`)}
+            </Flex>
+          )
+        })(),
```
Comment (fribbels) [inline src/components/RelicModal.tsx:223]: Why is this change needed? It looks like its defining an inline function then just calling it immediately?

File: src/components/importerTab/LoadDataSubmenu.tsx [hunk 3]
```diff
@@ -110,10 +112,10 @@ export function LoadDataSubmenu() {
       <Flex style={{ minHeight: 100 }}>
         <Flex vertical gap={10} style={{ display: currentStage >= 1 ? 'flex' : 'none' }}>
           <Text>
-            {`File contains ${currentSave.relics.length} relics and ${currentSave.characters.length} characters. Replace your current data with the uploaded data?`}
+            {t('stage2.label', { reliccount: currentSave.relics.length, charactercount: currentSave.characters.length })}
           </Text>
           <Button style={{ width: importerTabButtonWidth }} icon={<ImportOutlined/>} type='primary' onClick={loadConfirmed} loading={loading2}>
```
Comment (fribbels) [inline src/components/importerTab/LoadDataSubmenu.tsx:117]: Do you think its worth keeping a commented version of translated strings in the code in English for readability? It looks really hard to navigate the codebase now. Something like: 

```
{t('stage2.label', { reliccount: currentSave.relics.length, charactercount: currentSave.characters.length })}
// Translation: {`File contains ${currentSave.relics.length} relics and ${currentSave.characters.length} characters. Replace your current data with the uploaded data?`}
```

File: public/locales/en/relicsTab.json [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      "noneselected": "Select a $t(common:relic, {\"count\": 1, \"length\": 0}) to locate",
+      "location": "Location - Row {{rowindex}} / Col {{columnindex}}"
+    },
+    "insightoptions": {
+      "buckets": "$t(common:relic, {\"count\": 1, \"length\": 1}) Insight: Buckets",
+      "top10": "$t(common:relic, {\"count\": 1, \"length\": 1}) Insight: Top 10"
+    },
+    "plotoptions": {
+      "plotall": "Show all $t(common:character, {\"count\": 12, \"length\": 0})",
+      "plotcustom": "Show custom $t(common:character, {\"count\": 12, \"length\": 0})"
```
Comment (fribbels) [inline public/locales/en/relicsTab.json:107]: What does the count/length stuff do here?

File: misc/i18n/generateTranslation.js [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+  var hash2 = 5381;
+  for (let i = 0; i < key.length; i += 2)
+  {
+    hash1 = Math.imul((hash1 << 5) + hash1, 1) ^ key.charCodeAt(i);
+    if (i === key.length - 1)
+      break;
+    hash2 = Math.imul((hash2 << 5) + hash2, 1) ^ key.charCodeAt(i + 1);
+  }
+  return Math.imul(hash1 + Math.imul(hash2, 1566083941), 1);
+}
```
Comment (fribbels) [inline misc/i18n/generateTranslation.js:262]: Im not following what this is doing here, need some documentation

File: src/components/RelicFilterBar.jsx [hunk 6]
```diff
@@ -237,51 +240,51 @@ export default function RelicFilterBar(props) {
...
-          <HeaderText>Equipped</HeaderText>
+          <HeaderText>{t('relicFilterBar.equipped')}</HeaderText>
           <FilterRow name='equippedBy' tags={equippedByData} flexBasis='15%'/>
         </Flex>
         <Flex vertical flex={0.4}>
-          <HeaderText>Clear</HeaderText>
+          <HeaderText>{t('clear')}</HeaderText>
           <Button icon={<ClearOutlined/>} onClick={clearClicked} style={{ flexGrow: 1, height: '100%' }}>
-            Clear all filters
+            {t('relicFilterBar.clearbutton')}
```
Comment (fribbels) [inline src/components/RelicFilterBar.jsx:265]: There's some inconsistent case naming going on here. I think we have to decide on one of the following and stick with it for consistency

* RelicFilterBar.ClearButton
* relicFilterBar.clearButton
* relicfilterbase.clearbutton

I have a slight preference for the first one

Comment (FI00ds) [reply src/components/RelicFilterBar.jsx:265]: yeah I was initially wanting to keep everything lower case with uppercase being reserved for i18n specific things (e.g. "keyWithCount" and "key") but that's proven to not really hold up so well, especially with the keys for stats. will probably move to the first option.

File: misc/i18n/generateTranslation.js (hunk 5 above)
Comment (FI00ds) [reply misc/i18n/generateTranslation.js:262]: I'll fully admit I have no clue on the what/why for this.
This is what the stargazer guys use to find the hash when it isn't given in the files from Dim's repo

File: src/components/importerTab/LoadDataSubmenu.tsx (hunk 3 above)
Comment (FI00ds) [reply src/components/importerTab/LoadDataSubmenu.tsx:117]: yeah that's probably not a bad idea

File: src/components/CharacterModal.jsx (hunk 1 above)
Comment (FI00ds) [reply src/components/CharacterModal.jsx:62]: mixed case is handled by a "capitalize" formatter, so the keys can be case insensitive

**Comment by fribbels** [comment]
Im getting a loading page like this for a few seconds, probably not ideal UX
image

**Comment by fribbels** [comment]
Is Stelle/Caelus supposed to be missing their name translations? 
image

**Comment by fribbels** [comment]
Loading DPS score profiles feels significantly slower than beta, I'm not sure what going on there

**Comment by fribbels** [comment]
We should probably just limit this to a couple languages for the initial launch, I think probably just start with French & possibly Chinese if we find a translator. There's probably going to be like a boy-who-cried-wolf effect if we put up all the languages but only have the names translated where people won't try it again.

File: src/components/optimizerTab/OptimizerSuggestionsModal.tsx [hunk 7]
```diff
@@ -457,3 +511,59 @@ export function ZeroResultSuggestionModal() {
     </Modal>
   )
 }
+
+type Request = {
```
Comment (fribbels) [inline src/components/optimizerTab/OptimizerSuggestionsModal.tsx:515]: Is this the same as `import { Form } from 'types/Form'`?

File: src/components/optimizerTab/Sidebar.jsx [hunk 8]
```diff
@@ -15,40 +15,39 @@ import { Utils } from 'lib/utils'
...
 
 const { Text } = Typography
 
-const computeEngineToDisplay = {
-  [COMPUTE_ENGINE_GPU_EXPERIMENTAL]: 'GPU acceleration: Enabled',
-  [COMPUTE_ENGINE_GPU_STABLE]: 'GPU acceleration: Enabled',
-  [COMPUTE_ENGINE_CPU]: 'GPU acceleration: Disabled',
-}
-
 function getGpuOptions(computeEngine) {
```
Comment (fribbels) [inline src/components/optimizerTab/Sidebar.jsx:26]: This seems broken 
image

File: src/components/optimizerTab/Sidebar.jsx [hunk 9]
```diff
@@ -277,26 +287,26 @@ function SidebarContent() {
...
+                    description={t('Sidebar.ControlsGroup.ResetConfirm.Description')}// 'All filters will be reset to their default values'
                     onConfirm={resetClicked}
-                    okText='Yes'
-                    cancelText='No'
+                    okText={t('Sidebar.ControlsGroup.ResetConfirm.Yes')}// 'Yes'
+                    cancelText={t('Sidebar.ControlsGroup.ResetConfirm.No')}// 'No'
                   >
                     <Button style={{ flex: 1 }}>
-                      Reset
+                      {t('Sidebar.ControlsGroup.Reset')/* Reset */}
```
Comment (fribbels) [inline src/components/optimizerTab/Sidebar.jsx:309]: Should these be under common?

File: src/lib/conditionals/character/Argenti.ts [hunk 10]
```diff
@@ -25,46 +29,45 @@ export default (e: Eidolon): CharacterConditional => {
...
     {
       formItem: 'switch',
       id: 'enemyHp50',
       name: 'enemyHp50',
-      text: 'Enemy HP ≤ 50% DMG boost',
-      title: 'Courage',
-      content: `Deals 15% more DMG to enemies whose HP percentage is 50% or less.`,
+      text: t('Content.1.text'),
+      title: t('Content.1.title'),
+      content: t('Content.1.content'),
```
Comment (fribbels) [inline src/lib/conditionals/character/Argenti.ts:41]: I feel like these should be tied to the id/name rather than order index. `Argenti.enemyHp50Text` or something. Otherwise switching the order of the content will suddenly mess up the ordering for all the conditionals.

File: src/lib/conditionals/character/Arlan.ts [hunk 11]
```diff
@@ -1,13 +1,17 @@
...
 import { CharacterConditional } from 'types/CharacterConditional'
 import { Form } from 'types/Form'
 import { ContentItem } from 'types/Conditionals'
 import { buffAbilityDmg } from 'lib/optimizer/calculateBuffs'
+import i18next from 'i18next'
+import { TsUtils } from 'lib/TsUtils'
 
 export default (e: Eidolon): CharacterConditional => {
+  /* @ts-expect-error ts can't resolve the type 'Type instantiation is excessively deep and possibly infinite' */
+  const t = i18next.getFixedT(null, 'conditionals', 'Characters.Arlan')
```
Comment (fribbels) [inline src/lib/conditionals/character/Arlan.ts:14]: IS there an alternative to getFixedT that doesn't run into this error? Seems a bit concerning

File: misc/i18n/generateTranslation.js [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+import TextMapFR from './TextMapFR.json' assert {type: 'json'}
+import TextMapID from './TextMapID.json' assert {type: 'json'}
+import TextMapJP from './TextMapJP.json' assert {type: 'json'}
+import TextMapKR from './TextMapKR.json' assert {type: 'json'}
+import TextMapPT from './TextMapPT.json' assert {type: 'json'}
+import TextMapRU from './TextMapRU.json' assert {type: 'json'}
+import TextMapTH from './TextMapTH.json' assert {type: 'json'}
+import TextMapVI from './TextMapVI.json' assert {type: 'json'}
+
+function precisionRound(number) {
```
Comment (fribbels) [inline misc/i18n/generateTranslation.js:26]: Can we TS this file and put some info on how to run it? I think the ideal end state is to be able to run the script with `npx tsx misc/i18n/generateTranslations.ts`

But I'm getting errors with that

```
import AvatarConfig from './AvatarConfig.json' assert {type: 'json'}
                                               ^^^^^^

SyntaxError: Unexpected identifier 'assert'
    at compileSourceTextModule (node:internal/modules/esm/utils:337:16)
    at ModuleLoader.moduleStrategy (node:internal/modules/esm/translators:166:18)
    at callTranslator (node:internal/modules/esm/loader:436:14)
    at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:442:30)
    at async ModuleJob._link (node:internal/modules/esm/module_job:106:19)

Node.js v22.6.0
```

**Comment by fribbels** [comment]
I've uploaded the 2.6v4 beta configs but I would prefer to have as little beta information in this repo as possible - so we need translation backups for beta content character / light cone / relic set names. These probably can just come out of gameData.json since that contains the minimum necessary to run. Otherwise we become a target for DMCA's and that's very unpleasant

File: src/components/GettingStartedTab.jsx [hunk 13]
```diff
@@ -32,285 +52,386 @@ export default function GettingStartedTab() {
...
-          <h3>
-            Option 1: Load a sample save
-          </h3>
-          <p>
-            If you would like to give the optimizer a try before doing any relic importing, use this to load a sample
-            save file and check out the features.
-          </p>
+          <Trans t={t} i18nKey='TryOut.Option 1'>
+            <h3>
+              Option 1: Load a sample save THIS SHOULDN'T BE VISIBLE
```
[the remark this answers is not in the document]
Comment (FI00ds) [reply src/components/GettingStartedTab.jsx:60]: yeah I left that more as a way for me to make sure the Trans component was working properly, so the "HIS SHOULDN'T BE VISIBLE" can be removed without issue, the rest of the text should probably stay though

File: src/components/RelicScorerTab.jsx [hunk 14]
```diff
@@ -474,13 +486,13 @@ function CharacterPreviewSelection(props) {
...
-              Import relics into optimizer
+              {t('ImportLabels.Relics')/* Import relics into optimizer */}
             </Dropdown.Button>
             <Button icon={<ExperimentOutlined/>} onClick={simulateClicked} style={{ width: 280 }}>
-              Simulate relics on another character
+              {t('SimulateRelics')/* Simulate relics on another character */}
             </Button>
             <Button icon={<LineChartOutlined/>} onClick={optimizeClicked} style={{ width: 228 }}>
-              Optimize character stats
+              {t('OptimizeOnCharacter')/* Optimize character stats */}
```
[the remark this answers is not in the document]
Comment (FI00ds) [reply src/components/RelicScorerTab.jsx:495]: the namespace is provided when the t function is created { t } = UseTranslation(namespace), ImportLabels is just a keyprefix to group the text for the 3 import modes

File: src/components/optimizerTab/OptimizerSuggestionsModal.tsx (hunk 7 above)
Comment (FI00ds) [reply src/components/optimizerTab/OptimizerSuggestionsModal.tsx:515]: oh probably (: didn't realise we had a type for the form

File: src/lib/conditionals/character/Arlan.ts (hunk 11 above)
Comment (FI00ds) [reply src/lib/conditionals/character/Arlan.ts:14]: I'm not exactly sure what's causing the error because I was sometimes even seeing it in other files that don't use getFixedT

File: src/lib/conditionals/character/Argenti.ts (hunk 10 above)
Comment (fribbels) [reply src/lib/conditionals/character/Argenti.ts:41]: I can take this refactor if you're sick of the JSONs by now. Just making sure the reasoning is sound

File: src/components/optimizerTab/Sidebar.jsx (hunk 9 above)
Comment (FI00ds) [reply src/components/optimizerTab/Sidebar.jsx:309]: my thought process was to try and avoid directly accessing the common namespace in the code with the exception of certain things like stats. That way any change to the displayed text can be done solely by editing the locale files instead of having to both change the key and add a new entry in the translation files

File: src/lib/conditionals/character/Argenti.ts (hunk 10 above)
Comment (FI00ds) [reply src/lib/conditionals/character/Argenti.ts:41]: yeah if you could do that that would be great

**Comment by fribbels** [comment]
I think the set strings are getting mixed up with translations 
image
